modules/security-aai/src/main/java/org/opencastproject/security/aai/api/AttributeMapper.java

```diff
@@ -0,0 +1,186 @@
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ *
+ */
+package org.opencastproject.security.aai.api;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.InitializingBean;
+import org.springframework.expression.Expression;
+import org.springframework.expression.ExpressionParser;
+import org.springframework.expression.spel.standard.SpelExpressionParser;
+import org.springframework.util.Assert;
+import org.springframework.util.CollectionUtils;
+
+import java.util.HashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+import javax.servlet.http.HttpServletRequest;
+/**
+ * Generic AAI Attribute mapper using Spring Expression language mappings.
+ *
+ */
+public class AttributeMapper implements InitializingBean {
+
+  /** The logging facility */
+  private static final Logger logger = LoggerFactory
+      .getLogger(AttributeMapper.class);
+
+  /** List of all attributes that should be fetched */
+  private List<String> aaiAttributes;
+
+  /**
+   * Map of List of expressions. Key is a mapping name, value a list of
+   * expressions
+   */
+  private Map<String, List<String>> attributeMap;
+
+  /** Use HTTP Header or environment attributes */
+  private boolean useHeader = true;
+
+  /** The delimiter for multivalue attributes */
+  private String multiValueDelimiter = ";";
+
+  public void afterPropertiesSet() throws Exception {
+    Assert.notNull(attributeMap, "attributeMap must be set");
+  }
+
+  /**
+   * Apply all expressions on the sourceAttributes
+   *
+   * @param sourceAttributes
+   *    Key is attribute name, value a list of split AAI attribute values
+   * @param mappingId
+   *    The mapping list to use
+   * @return
+   */
+  public List<String> getMappedAttributes(
+      Map<String, List<String>> sourceAttributes, String mappingId) {
+    Set<String> mappedAttributes = new LinkedHashSet<String>();
+    ExpressionParser parser = new SpelExpressionParser();
+
+    List<String> expressions = attributeMap.get(mappingId);
+
+    if (expressions == null) {
+      logger.error("No mapping for \"" + mappingId
+          + "\" specified. Did you forget to configure a <util:map id=\""
+          + mappingId + "\" ...?");
+      throw new IllegalArgumentException("No mapping for \"" + mappingId
+          + "\" specified. Did you forget to configure a <util:map id=\""
+          + mappingId + "\" ...?");
+    }
+
+    for (String expression : expressions) {
+      Expression exp = null;
+      try {
+        exp = parser.parseExpression(expression);
+        String res = (String) exp.getValue(sourceAttributes);
+        logger.debug("Mapping " + exp.getExpressionString() + " -> " + res);
```

Comment:
This kind of change applies to a number of log messages in this patch:

```suggestion
        logger.debug("Mapping {} to {}", exp.getExpressionString(), res);
```